rfc-135-updating-routes-in-router.md

```diff
@@ -0,0 +1,53 @@
+
+The current implementation of updating routes has led to scaling issues. When scaling Router, Router API takes [5 minutes][] to get an accurate router node list. If Router API fails to update a Router node (i.e. one that is being brought up/down), the API also [fails to reach subsequent nodes in the list][].
+
+We should therefore rethink how these routes are updated within Router instances and to come up with another way of managing the update process.
+
+[5 minutes]: https://github.com/alphagov/govuk-puppet/blob/84689ae2c36bbc7740da6a8d70c7f4e04d0e2357/modules/govuk/manifests/apps/router_api.pp#L110
+[fails to reach subsequent nodes in the list]: https://github.com/alphagov/router-api/blob/2aa74d402624837c87fef136084ae66633e57d46/lib/router_reloader.rb#L46
+
+## Proposal
+
+Router will keep its in-memory cache of routes up-to-date via a two stage process:
+
+  1. Router will perform a simple query against MongoDB returning the state of the database with the current set of routes; ideally this will be a query which contains a checksum or timestamp of last update which can be used by Router instances to check whether their local copy of routes is up-to-date.
+  2. If the check against MongoDB concludes that Router needs to update its in-memory cache of routes, then Router will self-action a reload of all routes, using the same mechanism as it currently does when it receives a call to its `/reload` endpoint.
+
+The polling of changes by Router against MongoDB will take place every few seconds in order to ensure that routes are updated quickly as and when they change, which keeps in line with the current approach of updating routes when comparing how fast route updates propagate through to Router instances. Individual Router instances will drift slightly out of sync with each other, but because we're polling every few seconds and updating as soon as we see changes there won't be any noticeable changes.
+
+The query used to poll MongoDB for changes should be quick to compute, taking advantage of any existing indexes on the database or aggregates already stored by MongoDB; for example, the [db.stats() method](https://docs.mongodb.com/manual/reference/method/db.stats/) could be a useful starting point. The details of the query aren't included the scope of this RFC however and it should be assumed that a suitably performant query can be made which allows for the polling of changes to routes.
+
+This proposal is relatively simple to implement compared to some of the others we have considered and will allow us to move at pace in decoupling Router API and Router. It also puts us in a good place to iterate the solution further by using the change streams feature in future, whether that's offered by MongoDB, DocumentDB or any other database technology that we decide to use for storing routes.
```

Comment:
Would be good to link to a definition, for people unfamiliar with change streams (me!):

```suggestion
This proposal is relatively simple to implement compared to some of the others we have considered and will allow us to move at pace in decoupling Router API and Router. It also puts us in a good place to iterate the solution further by using [the change streams feature](#using-mongodb-change-streams-to-notify-router-instances-of-route-changes) in future, whether that's offered by MongoDB, DocumentDB or any other database technology that we decide to use for storing routes.
```